File: src/utils/CubicBezierCurve.ts
import type { Point2D, CubicBezierCurve } from '../types/geometry';
import { RectangleIntersection } from './RectangleIntersection';

export class CubicBezierCurveImpl implements CubicBezierCurve {
  private P0: Point2D;
  private P1: Point2D;
  private P2: Point2D;
  private P3: Point2D;
  private pathX: number[];
  private pathY: number[];
  private arcLength: number = -1;

  constructor(P0: Point2D, P1: Point2D, P2: Point2D, P3: Point2D) {
    this.P0 = P0;
    this.P1 = P1;
    this.P2 = P2;
    this.P3 = P3;

    const cubeTermX = P3.x - 3 * P2.x + 3 * P1.x - P0.x;
    const cubeTermY = P3.y - 3 * P2.y + 3 * P1.y - P0.y;
    const squareTermX = 3 * P2.x - 6 * P1.x + 3 * P0.x;
    const squareTermY = 3 * P2.y - 6 * P1.y + 3 * P0.y;
    const proportionalTermX = 3 * P1.x - 3 * P0.x;
    const proportionalTermY = 3 * P1.y - 3 * P0.y;
    const constantTermX = P0.x;
    const constantTermY = P0.y;

    this.pathX = [cubeTermX, squareTermX, proportionalTermX, constantTermX];
    this.pathY = [cubeTermY, squareTermY, proportionalTermY, constantTermY];
  }

  compute(t: number): Point2D {
    const mt = 1 - t;
    const mt2 = mt * mt;
    const mt3 = mt2 * mt;
    const t2 = t * t;
    const t3 = t2 * t;

    return {
      x: mt3 * this.P0.x + 3 * mt2 * t * this.P1.x + 3 * mt * t2 * this.P2.x + t3 * this.P3.x,
      y: mt3 * this.P0.y + 3 * mt2 * t * this.P1.y + 3 * mt * t2 * this.P2.y + t3 * this.P3.y
    };
  }

  computePathPrimes(t: number): Point2D {
    return {
      x: this.evaluatePolynomialDerivative(this.pathX, t),
      y: this.evaluatePolynomialDerivative(this.pathY, t)
    };
  }

  theta(t: number): number {
    const nabla = this.computePathPrimes(t);
    return Math.atan2(nabla.y, nabla.x);
  }

  pathInversion(point: Point2D): number {
    // Simple approximation - in practice you'd want a more sophisticated root finding algorithm
    let bestT = 0.5;
    let bestDistance = Number.MAX_VALUE;
    
    for (let t = 0; t <= 1; t += 0.01) {
      const computed = this.compute(t);
      const distance = Math.sqrt(
        Math.pow(computed.x - point.x, 2) + 
        Math.pow(computed.y - point.y, 2)
      );
      
      if (distance < bestDistance) {
        bestDistance = distance;
        bestT = t;
      }
    }
    
    return bestT;
  }

  getArcLength(): number {
    if (this.arcLength === -1) {
      // Simple approximation - can be improved if needed
      const steps = 100;
      let length = 0;
      let prevPoint = this.compute(0);

      for (let i = 1; i <= steps; i++) {
        const t = i / steps;
        const point = this.compute(t);
        length += Math.sqrt(
          Math.pow(point.x - prevPoint.x, 2) + 
          Math.pow(point.y - prevPoint.y, 2)
        );
        prevPoint = point;
      }
      this.arcLength = length;
    }
    return this.arcLength;
  }

  intersectionWeight(
    isBlue: boolean,
    theta: number,
    boundaryTolerance: number,
    submersibleTolerance: number,
    robotWidth: number,
    robotHeight: number
  ): number {
    let a = 0, b = 0, c = 0, d = 0, e = 0;
    
    for (let i = 0; i < 100; i++) {
      const center = this.compute(i / 100);
      a = this.collidesWithSubmersible(theta, center, submersibleTolerance, robotWidth, robotHeight);
      b = this.collidesWithAllianceBoundary(theta, center, boundaryTolerance, robotWidth, robotHeight);
      
      if (!isBlue) {
        c = this.collidesWithOuterBoundaryRed(theta, center, boundaryTolerance, robotWidth, robotHeight);
      } else {
        c = this.collidesWithOuterBoundaryBlue(theta, center, boundaryTolerance, robotWidth, robotHeight);
      }

      d = this.collidesWithLeftBoundary(theta, center, robotWidth, robotHeight, boundaryTolerance);
      e = this.collidesWithRightBoundary(theta, center, boundaryTolerance, robotWidth, robotHeight);
    }

    return 20 * (a + b + c + d + e) / 100;
  }

  private collidesWithSubmersible(
    theta: number,
    center: Point2D,
    submersibleTolerance: number,
    robotWidth: number,
    robotHeight: number
  ): number {
    const robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
    const submersible = RectangleIntersection.createRotatedRectangle(
      { x: 72, y: 72 },
      27.5 + 2 * submersibleTolerance,
      42.75 + 2 * submersibleTolerance,
      0
    );
    return RectangleIntersection.getMinimumSeparatingWidth(robot, submersible);
  }

  private collidesWithAllianceBoundary(
    theta: number,
    center: Point2D,
    boundaryTolerance: number,
    robotWidth: number,
    robotHeight: number
  ): number {
    const robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
    const boundary = RectangleIntersection.createRotatedRectangle(
      { x: 72, y: 72 },
      144,
      2 * boundaryTolerance,
      0
    );
    return RectangleIntersection.getMinimumSeparatingWidth(robot, boundary);
  }

  private collidesWithOuterBoundaryBlue(
    theta: number,
    center: Point2D,
    boundaryTolerance: number,
    robotWidth: number,
    robotHeight: number
  ): number {
    const robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
    const boundary = RectangleIntersection.createRotatedRectangle(
      { x: 0, y: 72 },
      144,
      boundaryTolerance,
      0
    );
    return RectangleIntersection.getMinimumSeparatingWidth(robot, boundary);
  }

  private collidesWithOuterBoundaryRed(
    theta: number,
    center: Point2D,
    boundaryTolerance: number,
    robotWidth: number,
    robotHeight: number
  ): number {
    const robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
    const boundary = RectangleIntersection.createRotatedRectangle(
      { x: 144, y: 72 },
      144,
      boundaryTolerance,
      0
    );
    return RectangleIntersection.getMinimumSeparatingWidth(robot, boundary);
  }

  private collidesWithRightBoundary(
    theta: number,
    center: Point2D,
    boundaryTolerance: number,
    robotWidth: number,
    robotHeight: number
  ): number {
    const robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
    const boundary = RectangleIntersection.createRotatedRectangle(
      { x: 72, y: 144 },
      boundaryTolerance,
      144,
      0
    );
    return RectangleIntersection.getMinimumSeparatingWidth(robot, boundary);
  }

  private collidesWithLeftBoundary(
    theta: number,
    center: Point2D,
    robotWidth: number,
    robotHeight: number,
    boundaryTolerance: number
  ): number {
    const robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
    const boundary = RectangleIntersection.createRotatedRectangle(
      { x: 72, y: 0 },
      boundaryTolerance,
      144,
      0
    );
    return RectangleIntersection.getMinimumSeparatingWidth(robot, boundary);
  }

  private evaluatePolynomial(coefficients: number[], t: number): number {
    let result = 0;
    for (let i = 0; i < coefficients.length; i++) {
      result += coefficients[i] * Math.pow(t, coefficients.length - 1 - i);
    }
    return result;
  }

  private evaluatePolynomialDerivative(coefficients: number[], t: number): number {
    let result = 0;
    for (let i = 0; i < coefficients.length - 1; i++) {
      result += (coefficients.length - 1 - i) * 
                coefficients[i] * 
                Math.pow(t, coefficients.length - 2 - i);
    }
    return result;
  }

  getP0(): Point2D { return this.P0; }
  getP1(): Point2D { return this.P1; }
  getP2(): Point2D { return this.P2; }
  getP3(): Point2D { return this.P3; }

  exportAsLine(degrees: number, color: string): Line {
    return {
      endPoint: { x: this.P3.x, y: this.P3.y, heading: "constant", degrees: degrees },
      controlPoints: [this.P1, this.P2],
      color: color
    };
  }
}

File: src/fpa/java/MathUtil/CubicBezierCurve.java
package MathUtil;

import org.apache.commons.math3.analysis.UnivariateFunction;
import org.apache.commons.math3.analysis.integration.SimpsonIntegrator;
import org.apache.commons.math3.analysis.polynomials.PolynomialFunction;
import org.apache.commons.math3.analysis.solvers.LaguerreSolver;
import org.apache.commons.math3.complex.Complex;
import org.apache.commons.math3.util.FastMath;

import java.awt.geom.Point2D;

public class CubicBezierCurve {
    private Point2D P_0;
    private Point2D P_1;
    private Point2D P_2;
    private Point2D P_3;
    PolynomialFunction pathX;
    PolynomialFunction pathY;

    private LaguerreSolver solver = new LaguerreSolver();
    private SimpsonIntegrator integrator = new SimpsonIntegrator();

    private double arcLength = -1;

    public CubicBezierCurve(Point2D P_0, Point2D P_1, Point2D P_2, Point2D P_3) {
        this.P_0 = P_0;
        this.P_1 = P_1;
        this.P_2 = P_2;
        this.P_3 = P_3;

        double cubeTermX = P_3.getX() - 3*P_2.getX() + 3*P_1.getX() - P_0.getX();
        double cubeTermY = P_3.getY() - 3*P_2.getY() + 3*P_1.getY() - P_0.getY();
        double squareTermX = 3*P_2.getX() - 6*P_1.getX() + 3*P_0.getX();
        double squareTermY = 3*P_2.getY() - 6*P_1.getY() + 3*P_0.getY();
        double proportionalTermX = 3*P_1.getX()-3*P_0.getX();
        double proportionalTermY = 3*P_1.getY()-3*P_0.getY();
        double constantTermX = P_0.getX();
        double constantTermY = P_0.getY();

        pathX = new PolynomialFunction(new double[] {cubeTermX, squareTermX, proportionalTermX, constantTermX});
        pathY = new PolynomialFunction(new double[] {cubeTermY, squareTermY, proportionalTermY, constantTermY});
    }

    public Point2D.Double compute(double t) {
        return new Point2D.Double(pathX.value(t), pathY.value(t));
    }

    public Point2D.Double computePathPrimes(double t) {
        return new Point2D.Double(pathX.polynomialDerivative().value(t), pathY.polynomialDerivative().value(t));
    }

    public double theta(double t) {
        Point2D.Double nabla = computePathPrimes(t);
        return FastMath.atan2(nabla.getY(),nabla.getX());
    }

    public double pathInversion(Point2D point) {
        Complex[] t_x = solver.solveAllComplex(pathX.getCoefficients(), 0.5, 300);
        Complex[] t_y = solver.solveAllComplex(pathY.getCoefficients(), 0.5, 300);

        double root = 0.5;

        for (Complex root_x : t_x) {
            for (Complex root_y : t_y) {
                if (root_x == root_y && root_x.getImaginary() == 0 && root_x.getReal() <= 1 && root_x.getReal() >= 0) {
                    root = root_x.getReal();
                }
            }
        }

        return root;
    }

    public double getArcLength() {
        if (arcLength == -1) {
            UnivariateFunction arcFunction = t -> computePathPrimes(t).distance(0,0);
            arcLength = integrator.integrate(200, arcFunction, 0, 1);
        }

        return arcLength;
    }

    public double intersectionWeight(boolean isBlue, double theta, double boundaryTolerance, double submersibleTolerance, double robotWidth, double robotHeight) {
        double a = 0,b = 0,c = 0,d = 0,e = 0;
        for (int i = 0; i < 100; i++) {
            Point2D.Double center = compute((double) i/100);
            a = collidesWithSubmersible(theta, center, submersibleTolerance, robotWidth, robotHeight);
            b = collidesWithAllianceBoundary(theta, center, boundaryTolerance, robotWidth, robotHeight);
            if (!isBlue) {
                c = collidesWithOuterBoundaryRed(theta, center, boundaryTolerance, robotWidth, robotHeight);
            } else {
                c = collidesWithOuterBoundaryBlue(theta, center, boundaryTolerance, robotWidth, robotHeight);
            }

            d = collidesWithLeftBoundary(theta, center, robotWidth, robotHeight, boundaryTolerance);

            e = collidesWithRightBoundary(theta, center, boundaryTolerance, robotWidth, robotHeight);
        }

        return 20*(a+b+c+d+e)/100;
    }

    public Point2D getP_0() {
        return P_0;
    }

    public Point2D getP_1() {
        return P_1;
    }

    public Point2D getP_2() {
        return P_2;
    }

    public Point2D getP_3() {
        return P_3;
    }

    private double collidesWithSubmersible(double theta, Point2D.Double center, double submersibleTolerance, double robotWidth, double robotHeight) {
        RectangleIntersection.Rectangle robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
        RectangleIntersection.Rectangle submersible = RectangleIntersection.createRotatedRectangle(new Point2D.Double(72, 72), 27.5 + 2*submersibleTolerance, 42.75 + 2*submersibleTolerance, 0);

        return RectangleIntersection.getMinimumSeparatingWidth(robot, submersible);
    }

    private double collidesWithAllianceBoundary(double theta, Point2D.Double center, double boundaryTolerance, double robotWidth, double robotHeight) {
        RectangleIntersection.Rectangle robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
        RectangleIntersection.Rectangle allianceBoundary = RectangleIntersection.createRotatedRectangle(new Point2D.Double(72, 72), 144, 2*boundaryTolerance, 0);
        return RectangleIntersection.getMinimumSeparatingWidth(robot,allianceBoundary);
    }

    private double collidesWithOuterBoundaryBlue(double theta, Point2D.Double center, double boundaryTolerance, double robotWidth, double robotHeight) {
        RectangleIntersection.Rectangle robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
        RectangleIntersection.Rectangle boundary = RectangleIntersection.createRotatedRectangle(new Point2D.Double(0, 72), 144, boundaryTolerance, 0);
        return RectangleIntersection.getMinimumSeparatingWidth(robot,boundary);
    }

    private double collidesWithOuterBoundaryRed(double theta, Point2D.Double center, double boundaryTolerance, double robotWidth, double robotHeight) {
        RectangleIntersection.Rectangle robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
        RectangleIntersection.Rectangle boundary = RectangleIntersection.createRotatedRectangle(new Point2D.Double(144, 72), 144, boundaryTolerance, 0);
        return RectangleIntersection.getMinimumSeparatingWidth(robot,boundary);
    }

    private double collidesWithRightBoundary(double theta, Point2D.Double center, double boundaryTolerance, double robotWidth, double robotHeight) {
        RectangleIntersection.Rectangle robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
        RectangleIntersection.Rectangle boundary = RectangleIntersection.createRotatedRectangle(new Point2D.Double(72, 144), boundaryTolerance, 144, 0);
        return RectangleIntersection.getMinimumSeparatingWidth(robot,boundary);
    }

    private double collidesWithLeftBoundary(double theta, Point2D.Double center, double robotWidth, double robotHeight, double boundaryTolerance) {
        RectangleIntersection.Rectangle robot = RectangleIntersection.createRotatedRectangle(center, robotWidth, robotHeight, theta);
        RectangleIntersection.Rectangle boundary = RectangleIntersection.createRotatedRectangle(new Point2D.Double(72, 0), boundaryTolerance, 144, 0);
        return RectangleIntersection.getMinimumSeparatingWidth(robot,boundary);
    }
}
